### PR TITLE
Emit TextInput onChangeText before onSelectionChange

### LINKED ISSUE #7740: TextInput onChangeText and onSelectionChange out of order
## Steps To Reproduce
https://user-images.githubusercontent.com/1106239/117464936-25535f80-af1f-11eb-8069-afaa710e596f.mov
Provide a detailed list of steps that reproduce the issue.

1. Open RNTester > TextInput component example > Text selection & cursor placement
2. Set breakpoints in onChangeText and onSelectionChange
3. Observe the order of events

## Expected Results
On iOS, Android, and macOS, the order of events is onChangeText -> onSelectionChange. On Windows, the order of events is reversed.

## Snack, code example, screenshot, or link to a repository:
Windows:

https://user-images.githubusercontent.com/1106239/117464104-58e1ba00-af1e-11eb-9794-812962b60073.mp4

Android:

https://user-images.githubusercontent.com/1106239/117464977-30a68b00-af1f-11eb-8f1c-ae0ceeb9fcf6.mov

iOS:

https://user-images.githubusercontent.com/1106239/117464915-1f5d7e80-af1f-11eb-8df5-0b74b4af1155.mov

### PR DESCRIPTION
**Pull request by rozele**
Previously, TextInputViewManager subscribed to two events to implement `onChangeText`:
1. TextChanging - used to increment a native counter to ensure JS changes to text do not overwrite native changes to text.
2. TextChanged - used to emit the `onChangeText` event and surprisingly also incremented the native counter.

Additionally, the view manager subscribed to SelectionChanged to emit the `onSelectionChange` event.

In XAML, the sequence of native events is as follows:
1. TextChanging
2. SelectionChanged
3. TextChanged

Since we emit the event that ultimately triggers `onChangeText` via the TextChanged native event, this means that we actually receive the cursor position change before the text change. This is different from other React Native platforms and makes it more difficult to implement behaviors based off the se`onSelectionChange` event.

One concern with this change is if there are any repercussions of eliminating the two phased approach to the native counter and firing of the event. Here is why I don't believe this will create any issues:

1. Previously, we incremented `m_nativeEventCount` twice, once in `TextChanging` and once in `TextChanged`. Reducing this to incrementing only once will not introduce any bugs, as JavaScript does not increment the `mostRecentEventCount` prop, so the only requirement is that a native change to the TextBox content increases the count.

2. Previously, we emitted the event in `TextChanged`. It appears that the `TextBox.Text()` property is already updated at the time that `TextChanging` is called, whether the `Text()` property was set externally or it was updated internally by a key event. The only difference between the two events is that `TextChanged` is called after the text is rendered (which may occur asynchronously from the actual text change). This is the reason we needed to introduce the dependency on `TextChanging` in the first place, as `TextChanging` is guaranteed to be called synchronously with the value change (and could not be interrupted by, e.g., an update from JS).

3. Is it a problem that JavaScript receives the text changed event before it actually renders? This is a good question, but I can't think of any scenarios where it's important that the TextBox has actually rendered the change before emitting the event, and since JS runs asynchronously from the UI thread, it's likely that by the time any UI changes / side effects triggered by JS make their way back to the UI thread, the text will have been rendered.

Fixes #7740

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7742)

### CODE REVIEW

**Comment by NickGerleman** [comment]
Any ideas on how we could test this sort of change?

**Comment by rozele** [comment]
> Any ideas on how we could test this sort of change?

It would probably be pretty trivial to create a test with an event counter, simulate typing a character and assert that the event count for onSelectionChange is less than the event count for onChange.

I'm not sure how WinAppDriver works, but if there's a way to simulate typing a character (that would also move the cursor forward), then I should be able to add an e2e test for this.

**Comment by NickGerleman** [comment]
We are able to simulate typing into a TextInput. See below for an example: https://github.com/microsoft/react-native-windows/blob/7a72b1e6e1630794067124d445105e0c3ba14f18/packages/e2e-test-app/test/LegacyLoginTest.test.ts#L69-L72

**Comment by rectified95** [comment]
Make sure this doesn't break the recently fixed scenario in https://github.com/microsoft/react-native-windows/pull/7658

**Comment by rozele** [comment]
> Make sure this doesn't break the recently fixed scenario in #7658

I don't believe it does for the reasons I documented. This really shouldn't change anything other than the order of events, since I don't believe we should have had any dependencies on the `onChange(Text)` event firing after rendering as opposed to synchronously with the text change.

**Comment by rozele** [comment]
@NickGerleman I added an E2E test for this. Please take another look.

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp [hunk 1]
```diff
@@ -177,9 +176,6 @@ void TextInputShadowNode::createView(const winrt::Microsoft::ReactNative::JSValu
 }
 
 void TextInputShadowNode::dispatchTextInputChangeEvent(winrt::hstring newText) {
-  if (!m_initialUpdateComplete) {
```
[the remark this answers is not in the document]
Comment (rozele) [reply vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:180]: I believe it prevents emitting an event on the initial load of, e.g., a value property.

**Review by NickGerleman — APPROVED** [review]

**Comment by NickGerleman** [comment]
@rozele you will have to run `yarn lint:fix` in the repo root to fix prettier formatting errors. If you have the eslint vscode extensions configured this formatting should be automatic on save.

**Pull request merged**